pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/PinotHelixResourceManager.java

```diff
@@ -1768,6 +1770,63 @@ public int reloadSegment(String tableNameWithType, String segmentName) {
     return numMessagesSent;
   }
 
+  /**
+   * Resets a segment by disabling and then enabling the segment
+   */
+  public void resetSegment(String tableNameWithType, String segmentName) {
+    IdealState idealState = getTableIdealState(tableNameWithType);
+    Preconditions.checkState(idealState != null, "Could not find ideal state for table: %s", tableNameWithType);
+    ExternalView externalView = getTableExternalView(tableNameWithType);
+    Preconditions.checkState(externalView != null, "Could not find external view for table: %s", tableNameWithType);
+    Set<String> instanceSet = idealState.getInstanceSet(segmentName);
+    Preconditions
+        .checkState(CollectionUtils.isNotEmpty(instanceSet), "Could not find segment: %s in ideal state for table: %s");
+    Map<String, String> externalViewStateMap = externalView.getStateMap(segmentName);
+    List<String> partitions = Lists.newArrayList(segmentName);
+    for (String instance : instanceSet) {
+      if (externalViewStateMap == null || SegmentStateModel.ERROR.equals(externalViewStateMap.get(instance))) {
+        _helixAdmin.resetPartition(_helixClusterName, instance, tableNameWithType, partitions);
+      } else {
+        _helixAdmin.enablePartition(false, _helixClusterName, instance, tableNameWithType, partitions);
+      }
```

Comment:
I remember we used to have this disable/enable (aka toggle)  in segments API before, we had a lot of problems with it. We used it for refreshing segments, and then we moved to use the refresh helix message instead.

Reply:
the problem with using SegmentRefreshMessage is that there's no way to bring about a change of state in the External View with that. A consuming segment in ERROR state will remain in ERROR state even after refreshed, unless we go through disable/enable partition

Reply:
Oh I know that. I am not saying that we use a message here. I am just pointing out that we had problems with the toggle API before. I forget the details, but it probably had to do with not waiting until EV stabilized

Reply:
yes, we might have to wait for EV to change/stabilize